rust/src/address.rs

```diff
@@ -0,0 +1,559 @@
+use super::*;
+use prelude::*;
+use bech32::ToBase32;
+
+// returns (Number represented, bytes read) if valid encoding
+// or None if decoding prematurely finished
+fn variable_nat_decode(bytes: &[u8]) -> Option<(u64, usize)> {
+    let mut output = 0u64;
+    let mut bytes_read = 0;
+    for byte in bytes {
+        output = (output << 7) | (byte & 0x7F) as u64;
+        bytes_read += 1;
+        if (byte & 0x80) == 0 {
+            return Some((output, bytes_read));
+        }
+    }
+    None
+}
+
+fn variable_nat_encode(mut num: u64) -> Vec<u8> {
+    let mut output = vec![num as u8 & 0x7F];
+    num /= 128;
+    while num > 0 {
+        output.push((num & 0x7F) as u8 | 0x80);
+        num /= 128;
+    }
+    output.reverse();
+    output
+}
+
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+enum AddrCredType {
+    Key(Keyhash),
+    Script(Scripthash),
+}
+
+#[wasm_bindgen]
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub struct AddrCred(AddrCredType);
+
+#[wasm_bindgen]
+impl AddrCred {
+    pub fn from_keyhash(hash: Keyhash) -> Self {
+        AddrCred(AddrCredType::Key(hash))
+    }
+
+    pub fn from_scripthash(hash: Scripthash) -> Self {
+        AddrCred(AddrCredType::Script(hash))
+    }
+
+    pub fn to_keyhash(&self) -> Option<Keyhash> {
+        match &self.0 {
+            AddrCredType::Key(hash) => Some(hash.clone()),
+            AddrCredType::Script(_) => None,
+        }
+    }
+
+    pub fn to_scripthash(&self) -> Option<Scripthash> {
+        match &self.0 {
+            AddrCredType::Key(_) => None,
+            AddrCredType::Script(hash) => Some(hash.clone()),
+        }
+    }
+
+    pub fn kind(&self) -> u8 {
+        match &self.0 {
+            AddrCredType::Key(_) => 0,
+            AddrCredType::Script(_) => 1,
+        }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match &self.0 {
+            AddrCredType::Key(hash) => hash.0.clone(),
+            AddrCredType::Script(hash) => hash.0.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+enum AddrType {
+    Base(BaseAddress),
+    Ptr(PointerAddress),
+    Enterprise(EnterpriseAddress),
+}
+
+#[wasm_bindgen]
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Address(AddrType);
+
+// to/from_bytes() are the raw encoding without a wrapping CBOR Bytes tag
```

Comment:
No sure if I understand this comment. `to_bytes` should be exactly the inverse to `bech32`.

Reply:
Within the context of the other CBOR serialization, we encode addresses as a CBOR byte string which has the Bytes major type length tag first. For to/from bytes I just made it the raw bytes (the ones you would use for the data portion of bech32 but no human readable portion).